File: packages/ap_common_plugin/android/src/main/kotlin/me/rainvisitor/ap_common_plugin/CoursePalette.kt
package me.rainvisitor.ap_common_plugin

import android.content.Context
import android.content.res.Configuration
import androidx.compose.ui.graphics.Color

/**
 * Reads the palette the Flutter side pushed via
 * `ApCommonPlugin.setCoursePalette`, falling back to the legacy
 * hardcoded Material 400 palette when nothing has been set yet.
 */
object CoursePalette {
    /** Legacy hardcoded palette — identical to the old hardcoded
     *  `courseColorPalette` that lived inside each widget. Used when
     *  the Flutter app hasn't pushed a palette yet (fresh install). */
    private val legacyColors = listOf(
        Color(0xFF5C6BC0),
        Color(0xFF26A69A),
        Color(0xFFEF5350),
        Color(0xFFAB47BC),
        Color(0xFF42A5F5),
        Color(0xFFFF7043),
        Color(0xFF66BB6A),
        Color(0xFFFFCA28),
        Color(0xFF8D6E63),
        Color(0xFF78909C),
        Color(0xFFEC407A),
        Color(0xFF7E57C2),
    )

    data class Resolved(
        val colors: List<Color>,
        val foreground: Color,
    )

    fun resolve(context: Context): Resolved {
        val prefs = context.getSharedPreferences(
            ApCommonPlugin.PREF_NAME,
            Context.MODE_PRIVATE,
        )
        val isDark =
            (context.resources.configuration.uiMode and
                Configuration.UI_MODE_NIGHT_MASK) ==
                Configuration.UI_MODE_NIGHT_YES

        val colorsKey =
            if (isDark)
                ApCommonPlugin.KEY_COURSE_PALETTE_DARK_COLORS
            else
                ApCommonPlugin.KEY_COURSE_PALETTE_COLORS
        val foregroundKey =
            if (isDark)
                ApCommonPlugin.KEY_COURSE_PALETTE_DARK_FOREGROUND
            else
                ApCommonPlugin.KEY_COURSE_PALETTE_FOREGROUND

        // Fall back to the light variant if dark hasn't been pushed.
        val rawColors = prefs.getString(colorsKey, null)
            ?: prefs.getString(
                ApCommonPlugin.KEY_COURSE_PALETTE_COLORS,
                null,
            )
        val rawForeground = prefs.getLong(foregroundKey, Long.MIN_VALUE)
            .takeIf { it != Long.MIN_VALUE }
            ?: prefs.getLong(
                ApCommonPlugin.KEY_COURSE_PALETTE_FOREGROUND,
                Long.MIN_VALUE,
            )

        val colors = rawColors
            ?.split(",")
            ?.mapNotNull { it.trim().toLongOrNull()?.let(::argbToColor) }
            ?.takeIf { it.isNotEmpty() }
            ?: legacyColors
        val foreground =
            if (rawForeground != Long.MIN_VALUE) argbToColor(rawForeground)
            else Color.White

        return Resolved(colors = colors, foreground = foreground)
    }

    private fun argbToColor(argb: Long): Color {
        val alpha = ((argb shr 24) and 0xFF).toInt()
        val red = ((argb shr 16) and 0xFF).toInt()
        val green = ((argb shr 8) and 0xFF).toInt()
        val blue = (argb and 0xFF).toInt()
        return Color(red = red, green = green, blue = blue, alpha = alpha)
    }
}

File: packages/ap_common_plugin/ios/CourseAppWidget/CoursePalette.swift
import SwiftUI
import WidgetKit

/// Reads the palette the Flutter side pushed via
/// `ApCommonPlugin.setCoursePalette`, falling back to the legacy
/// hardcoded Material 400 palette when no palette has been set yet.
enum CoursePalette {
    private static let keyId = "course_palette_id"
    private static let keyColors = "course_palette_colors"
    private static let keyForeground = "course_palette_foreground"
    private static let keyDarkColors = "course_palette_dark_colors"
    private static let keyDarkForeground = "course_palette_dark_foreground"

    /// Legacy Material 400 palette — matches the previous hardcoded
    /// list so existing installs keep the same visual until the user
    /// opens the app and pushes a new palette.
    private static let legacyColors: [Color] = [
        Color(red: 0.36, green: 0.42, blue: 0.75), // Indigo
        Color(red: 0.15, green: 0.65, blue: 0.60), // Teal
        Color(red: 0.94, green: 0.33, blue: 0.31), // Red
        Color(red: 0.67, green: 0.28, blue: 0.74), // Purple
        Color(red: 0.26, green: 0.65, blue: 0.96), // Blue
        Color(red: 1.00, green: 0.44, blue: 0.26), // Deep Orange
        Color(red: 0.40, green: 0.73, blue: 0.42), // Green
        Color(red: 1.00, green: 0.79, blue: 0.16), // Amber
        Color(red: 0.55, green: 0.43, blue: 0.39), // Brown
        Color(red: 0.47, green: 0.56, blue: 0.61), // Blue Grey
        Color(red: 0.93, green: 0.25, blue: 0.48), // Pink
        Color(red: 0.49, green: 0.34, blue: 0.76), // Deep Purple
    ]

    struct Resolved {
        let colors: [Color]
        let foreground: Color
    }

    /// Resolves the palette for the given [colorScheme], picking dark
    /// variants when available.
    static func resolve(for colorScheme: ColorScheme) -> Resolved {
        guard let groupDefaults = UserDefaults(suiteName: appGroupId) else {
            return Resolved(colors: legacyColors, foreground: .white)
        }
        let isDark = colorScheme == .dark

        let colorsKey = isDark ? keyDarkColors : keyColors
        let foregroundKey = isDark ? keyDarkForeground : keyForeground

        let rawColors = groupDefaults.array(forKey: colorsKey)
            ?? groupDefaults.array(forKey: keyColors)
        let rawForeground = groupDefaults.object(forKey: foregroundKey)
            ?? groupDefaults.object(forKey: keyForeground)

        let colors: [Color]
        if let raw = rawColors as? [Int64], !raw.isEmpty {
            colors = raw.map { argbToColor($0) }
        } else if let raw = rawColors as? [Int], !raw.isEmpty {
            colors = raw.map { argbToColor(Int64($0)) }
        } else if let raw = rawColors as? [NSNumber], !raw.isEmpty {
            colors = raw.map { argbToColor($0.int64Value) }
        } else {
            colors = legacyColors
        }

        let foreground: Color
        if let number = rawForeground as? NSNumber {
            foreground = argbToColor(number.int64Value)
        } else {
            foreground = .white
        }

        return Resolved(colors: colors, foreground: foreground)
    }

    private static func argbToColor(_ argb: Int64) -> Color {
        let alpha = Double((argb >> 24) & 0xFF) / 255.0
        let red = Double((argb >> 16) & 0xFF) / 255.0
        let green = Double((argb >> 8) & 0xFF) / 255.0
        let blue = Double(argb & 0xFF) / 255.0
        return Color(.sRGB, red: red, green: green, blue: blue, opacity: alpha)
    }
}
